### PR TITLE
added first attempt at solving assigment 3

### PR DESCRIPTION
**Pull request by Sephora-M**
Hi, 

Sorry about this last last minute PR... It has been a crazy couple of days!
I am aware that there are still gaps to be filled, mainly testing and documenting. I'll do that asap I promise, but right now I need to sleep ^^

### CODE REVIEW

**Comment by Sephora-M** [comment]
Hi,

So I've added the test and some documentation. I realise there is so much more that can be improved (such as more tests (in particular edge cases), improvement of the Trie data structure (could contain method to print all the words), and so on and so forth. But I could keep going for hours so I'll wait for your feedback Izabela to know where should I focus my efforts.

Thanks!

File: Sephora-M/assignment3/LetterGrid.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+public class LetterGrid {
+	char[][] grid;
+	int X;
+	int Y;
+	
+	public LetterGrid(char[][] grid) {
+		if (grid == null || grid[0].length == 0)
+			throw new IllegalArgumentException("Cannot instatiate an empty or a null grid.");
+		this.grid = grid;
+		Y = grid.length;
```
Comment (imorlowska) [inline Sephora-M/assignment3/LetterGrid.java]: optional: I find using X and Y can be a bit confusing sometimes when dealing with tables. It might be easier to read if you used rowCount and columnCount, but X and Y is fine.

File: Sephora-M/assignment3/LetterGrid.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.Set;
+
+/**Grid of letters. Contains a method that, given a dictionnary represented as a Trie, allows to find 
+ * all the words that can be formed in the grid. 
+ * 
+ * @author Sephora-M
+ */
+public class LetterGrid {
+	char[][] grid;
+	int X;
```
Comment (imorlowska) [inline Sephora-M/assignment3/LetterGrid.java]: All three of these can be `private final`, correct? :)

File: Sephora-M/assignment3/LetterGrid.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		Y = grid.length;
+		X = grid[0].length;
+	}
+	
+	/**
+	 * Given a Trie object, returns a set of all the words in the Trie that can be formed in the grid 
+	 * @param dict a Trie, the dictionary
+	 * @return set of all the words from the dictionary that can be from in the grid
+	 * 
+	 * Note: the idea is to start from each letter in the grid and perform a DFS-like search. To optimize
```
Comment (imorlowska) [inline Sephora-M/assignment3/LetterGrid.java]: nit: I'd put this between the method description and the `@param` line.

File: Sephora-M/assignment3/LetterGrid.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	 * depthMap, which tells us when we visited each node. 
+	 */
+	public Set<String> dictionnaryWords(Trie dict){
+		HashSet<String> foundWords = new HashSet<String>();
+
+		for(int x = 0; x < X; x++) {
+			for(int y = 0; y < Y; y++) {
+				int[][] depthMap = new int[Y][X];
+				int currentDepth = 1;
+				depthMap[y][x] = currentDepth;
```
Comment (imorlowska) [inline Sephora-M/assignment3/LetterGrid.java]: If you do a DFS recursively from each node, mark a node visited when you enter it in searchWords (if it hasn't been visited yet) and then unmark it as visited at the end of searchWords, then you can just use one boolean[][] array here, and reuse it instead of doing deep copies. This would save a lot of space for even small grids.

File: Sephora-M/assignment3/LetterGrid.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+		
+		return foundWords;
+	}
+	
+    private void searchWords(Point letter, int currentDepth, int[][] depthMap, Trie dict,
+			HashSet<String> foundWords, String currentPrefix) {
+    	
+    	ArrayList<Point> neighbors = findNeighbors(letter, currentDepth, depthMap);
+    	
+    	if (neighbors.size() == 0)
```
Comment (imorlowska) [inline Sephora-M/assignment3/LetterGrid.java]: optional: No need for checking for `size == 0`, if the list is empty then the `for` statement underneath just won't execute its content.

File: Sephora-M/assignment3/LetterGrid.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    	for (Point n : neighbors) {
+    		String newPrefix = currentPrefix + getLetter(n);
+    		if (dict.isPrefix(newPrefix)){
+    			int[][] newDepthMap = deepCopy(depthMap);
+    			newDepthMap[n.y][n.x] = currentDepth + 1;
+    			if (dict.isWord(newPrefix))
+    				foundWords.add(newPrefix);
+    			searchWords(n, currentDepth +1, newDepthMap, dict, foundWords, newPrefix);
+    		}
+    	}
```
Comment (imorlowska) [inline Sephora-M/assignment3/LetterGrid.java]: nit: please fix indentation and extra empty lines.

File: Sephora-M/assignment3/LetterGrid.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        for (int x = 0; x < X; x++) {
+          for (int y = 0; y < Y; y++) {
+        		  copy[y][x] = depthMap[y][x];
+          }
+        }
+        
+        return copy;
+      }
+    
+	private ArrayList<Point> findNeighbors(Point p, int currentDepth, int[][] depthMap){
```
Comment (imorlowska) [inline Sephora-M/assignment3/LetterGrid.java]: Very nice! How about writing a few unit tests for this method? When I was writing my solution to this problem I actually had a bug in this method and couldn't find it for good 10 minutes. :) It's important to unit test methods that are either algorithmically complex or easy to do a typo in (accidentally writing a '+' instead of a '-' for example).

File: Sephora-M/assignment3/LetterGridTest.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import org.junit.jupiter.api.Test;
+
+/**
+ * Note: should improve test by using setup and by testing edge cases...
+ * @author Sephora-M
+ *
+ */
+class LetterGridTest {
+	
+	@Test
```
Comment (imorlowska) [inline Sephora-M/assignment3/LetterGridTest.java]: I love how you tested the `isWord` method! :100: 
How about also testing the `isPrefix` method?

File: Sephora-M/assignment3/LetterGridTest.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	void testNotInDictionnary() {
+		String[] words = {"CAR","CARDS","CART","CAT"};
+		Trie dictionnary = new Trie(words);
+
+		assertFalse(dictionnary.isWord("CATS"));
+		assertFalse(dictionnary.isWord(""));
+	}
+	
+	@Test
```
Comment (imorlowska) [inline Sephora-M/assignment3/LetterGridTest.java]: Maybe a few more edge-cases if you have the time: empty grid, empty dictionary, words that would work if re-using letters was allowed, words that would work if skipping cells was allowed, words on diagonals etc.

File: Sephora-M/assignment3/LetterGrid.java (hunk 2 above)
Comment (Sephora-M) [reply Sephora-M/assignment3/LetterGrid.java]: correct! fixed :)

File: Sephora-M/assignment3/LetterGrid.java (hunk 1 above)
Comment (Sephora-M) [reply Sephora-M/assignment3/LetterGrid.java]: yea, I agree.. it actually caused me a bit of a headache because. I will keep in mind for next time!

File: Sephora-M/assignment3/LetterGrid.java (hunk 3 above)
Comment (Sephora-M) [reply Sephora-M/assignment3/LetterGrid.java]: ok :)

File: Sephora-M/assignment3/LetterGrid.java (hunk 4 above)
Comment (Sephora-M) [reply Sephora-M/assignment3/LetterGrid.java]: Ahh I was scratching my head to find this solution.. I didn't think of simply setting it back to unvisited at the end of searchWord.. I love this solution, I've adapted the code accordingly, thanks a lot!

File: Sephora-M/assignment3/LetterGrid.java (hunk 5 above)
Comment (Sephora-M) [reply Sephora-M/assignment3/LetterGrid.java]: true that!
